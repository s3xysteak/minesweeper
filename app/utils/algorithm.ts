export function shuffle<T>(arr: readonly T[], rand: (start: number, end: number) => number): T[] {
  const val = [...arr]
  const len = val.length
  for (let i = len - 1; i >= 0; i--) {
    const j = rand(0, i)
    ;[val[j], val[i]] = [val[i]!, val[j]!]
  }
  return val
}

export function shuffle2D<T>(arr: readonly T[][], rand: (start: number, end: number) => number): T[][] {
  let cursor = 0
  const newArr = shuffle(arr.flat(), rand)

  return arr.map(list => newArr.slice(cursor, cursor += list.length))
}
