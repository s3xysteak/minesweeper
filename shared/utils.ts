export function rollSeed() {
  return Math.floor(Math.random() * 0xFFFFFFFF)
}
