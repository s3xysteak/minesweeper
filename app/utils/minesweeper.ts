import type rand from 'twistrand'
import type { MineBlockType } from '~/components/types'
import { shuffle2D } from './algorithm'

export interface MinesweeperOptions {
  height: number
  width: number
  mineCount: number
}

export function createMines(mt: ReturnType<typeof rand>, opts: MinesweeperOptions): (MineBlockType & { col: number, row: number })[][] {
  const {
    height,
    width,
    mineCount,
  } = opts

  let _mineSpawnCount = 0

  const mineTable = shuffle2D(
    Array.from({ length: height }, () => Array.from({ length: width }, () => mineCount > _mineSpawnCount++)),
    (min, max) => mt.randomInt(min, max),
  )

  const offset = [
    { x: -1, y: -1 },
    { x: 0, y: -1 },
    { x: 1, y: -1 },
    { x: -1, y: 0 },
    { x: 1, y: 0 },
    { x: -1, y: 1 },
    { x: 0, y: 1 },
    { x: 1, y: 1 },
  ]

  return mineTable.map((row, rowIndex) =>
    row.map((isMine, colIndex) => (<MineBlockType & { col: number, row: number }>{
      col: colIndex,
      row: rowIndex,
      type: isMine ? 'bomb' : 'normal',
      revealed: false,
      clickable: 'ok',
      bombsAround: offset.reduce((acc, val) => {
        const { x, y } = val
        return mineTable?.[rowIndex + x]?.[colIndex + y] ? acc + 1 : acc
      }, 0),
    })))
}
