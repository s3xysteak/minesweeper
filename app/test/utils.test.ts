import rand from 'twistrand'
import { describe, expect, it } from 'vitest'
import { shuffle, shuffle2D } from '../utils/algorithm'
import { createMines } from '../utils/minesweeper'

describe('algorithm', () => {
  const mt = rand(12345)

  it('shuffle', () => {
    expect(
      shuffle(
        Array.from({ length: 10 }, (_, i) => i),
        (min, max) => mt.randomInt(min, max),
      ),
    ).toMatchSnapshot()
  })

  it('shuffle2D', () => {
    expect(
      shuffle2D(
        Array.from({ length: 10 }, (_, i) => [i]),
        (min, max) => mt.randomInt(min, max),
      ),
    ).toMatchSnapshot()

    expect(
      shuffle2D(
        [[1], [2, 3], [4, 5, 6], [7, 8, 9]],
        (min, max) => mt.randomInt(min, max),
      ),
    ).toMatchSnapshot()
  })
})

describe('minesweeper', () => {
  it('createMines', () => {
    expect(
      createMines(rand(12345), {
        height: 4,
        width: 3,
        mineCount: 5,
      }),
    ).toMatchSnapshot()
  })
})
