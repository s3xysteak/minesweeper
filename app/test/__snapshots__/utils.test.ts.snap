// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`algorithm > shuffle 1`] = `
[
  4,
  7,
  3,
  5,
  0,
  8,
  9,
  2,
  1,
  6,
]
`;

exports[`algorithm > shuffle2D 1`] = `
[
  [
    5,
  ],
  [
    1,
  ],
  [
    7,
  ],
  [
    4,
  ],
  [
    9,
  ],
  [
    2,
  ],
  [
    0,
  ],
  [
    6,
  ],
  [
    3,
  ],
  [
    8,
  ],
]
`;

exports[`algorithm > shuffle2D 2`] = `
[
  [
    9,
  ],
  [
    3,
    4,
  ],
  [
    8,
    6,
    5,
  ],
  [
    2,
    1,
    7,
  ],
]
`;

exports[`minesweeper > createMines 1`] = `
[
  [
    {
      "bombsAround": 0,
      "clickable": "ok",
      "col": 0,
      "revealed": false,
      "row": 0,
      "type": "normal",
    },
    {
      "bombsAround": 2,
      "clickable": "ok",
      "col": 1,
      "revealed": false,
      "row": 0,
      "type": "normal",
    },
    {
      "bombsAround": 1,
      "clickable": "ok",
      "col": 2,
      "revealed": false,
      "row": 0,
      "type": "bomb",
    },
  ],
  [
    {
      "bombsAround": 1,
      "clickable": "ok",
      "col": 0,
      "revealed": false,
      "row": 1,
      "type": "normal",
    },
    {
      "bombsAround": 3,
      "clickable": "ok",
      "col": 1,
      "revealed": false,
      "row": 1,
      "type": "normal",
    },
    {
      "bombsAround": 1,
      "clickable": "ok",
      "col": 2,
      "revealed": false,
      "row": 1,
      "type": "bomb",
    },
  ],
  [
    {
      "bombsAround": 2,
      "clickable": "ok",
      "col": 0,
      "revealed": false,
      "row": 2,
      "type": "bomb",
    },
    {
      "bombsAround": 4,
      "clickable": "ok",
      "col": 1,
      "revealed": false,
      "row": 2,
      "type": "normal",
    },
    {
      "bombsAround": 2,
      "clickable": "ok",
      "col": 2,
      "revealed": false,
      "row": 2,
      "type": "normal",
    },
  ],
  [
    {
      "bombsAround": 2,
      "clickable": "ok",
      "col": 0,
      "revealed": false,
      "row": 3,
      "type": "bomb",
    },
    {
      "bombsAround": 2,
      "clickable": "ok",
      "col": 1,
      "revealed": false,
      "row": 3,
      "type": "bomb",
    },
    {
      "bombsAround": 1,
      "clickable": "ok",
      "col": 2,
      "revealed": false,
      "row": 3,
      "type": "normal",
    },
  ],
]
`;
